packager/packager.cc

```diff
@@ -1542,6 +1556,83 @@ unique_ptr<PackageInfoImpl> createAndPopulatePackageInfo(core::GlobalState &gs,
     return info;
 }
 
+// Metadata for Tarjan's algorithm
+struct ComputeSCCsMetadata {
+    int nextIndex = 1;
+    int nextSCCId = 0;
+    // A given package's index in the DFS traversal; ie. when it was first visited
+    UnorderedMap<core::packages::MangledName, int> index;
+    // The lowest index reachable from a given package (in the same SCC) by following any number of tree edges
+    // and at most one back/cross edge
+    UnorderedMap<core::packages::MangledName, int> lowLink;
+    UnorderedMap<core::packages::MangledName, bool> onStack;
+    std::vector<core::packages::MangledName> stack;
+};
+
+// DFS traversal for Tarjan's algorithm starting from pkgName, along with keeping track of some metadata needed for
+// detecting SCCs.
+void strongConnect(core::GlobalState &gs, ComputeSCCsMetadata &metadata, core::packages::MangledName pkgName) {
+    if (!gs.packageDB().getPackageInfo(pkgName).exists()) {
+        return;
+    }
```

Comment:
Is this defensive? IF not, when can this happen? Does this only happen because our `allPackages` loop below implicitly loops over the NONE_PKG?

Would it be better to change this to an `ENFORCE`? or are there cases where we store NONE_PKG in the imports of a package? If so, it would be nice to document what those cases are.

Reply:
```suggestion
    // This is to handle the case where the user imports a package that doesn't exist.
    if (!gs.packageDB().getPackageInfo(pkgName).exists()) {
        return;
    }
```

In the future, imports will store a PackageRef, not a MangledName, and there will not be a PackageRef if the import target does not exist, so this case will go away. I want to leave a comment to capture that we can clean this up in the future.